File: backend/libs/money/src/main/java/de/bennyboer/kicherkrabbe/money/Money.java
package de.bennyboer.kicherkrabbe.money;

import lombok.AllArgsConstructor;
import lombok.Value;

import static lombok.AccessLevel.PRIVATE;

@Value
@AllArgsConstructor(access = PRIVATE)
public class Money {

    long amount;

    Currency currency;

    public static Money of(long amount, Currency currency) {
        return new Money(amount, currency);
    }

    public static Money euro(long amount) {
        return new Money(amount, Currency.euro());
    }

    public boolean isPositiveOrZero() {
        return amount >= 0;
    }

}

File: frontend/management/src/app/modules/admin/modules/offers/model/money.ts
import { validateProps } from '@kicherkrabbe/shared';

export class Money {
  readonly amount: number;
  readonly currency: string;

  private constructor(props: { amount: number; currency: string }) {
    validateProps(props);

    this.amount = props.amount;
    this.currency = props.currency;
  }

  static of(props: { amount: number; currency: string }): Money {
    return new Money({
      amount: props.amount,
      currency: props.currency,
    });
  }

  static euro(amountInCents: number): Money {
    return new Money({
      amount: amountInCents,
      currency: 'EUR',
    });
  }

  toDisplayString(): string {
    const euros = (this.amount / 100).toFixed(2).replace('.', ',');
    return `${euros} \u20AC`;
  }
}
